Fixed array out of bounds error

delpantheone opened the pull request:
This is an attempt to fix the already know "array out of bounds" error.

Host: Dell Inspiron 15 3520
OS: Fedora Linux 42 (Workstation Edition) x86_64
Kernel: Linux 6.14.4-300.fc42.x86_64

delpantheone committed:
Fixed array out of bounds error

wiktorek140 commented on os_dep/linux/ioctl_cfg80211.c:
```diff
@@ -1567,11 +1567,18 @@ static int rtw_cfg80211_set_encryption(struct net_device *dev, struct ieee_param
-						_rtw_memcpy(psta->dot11tkiprxmickey.skey, &(param->u.crypt.key[24]), 8);
-						padapter->securitypriv.busetkipkey = _FALSE;
-					}
+          if (strcmp(param->u.crypt.alg, "TKIP") == 0) { /* set mic key */
+              if (param->u.crypt.key_len >= 32) {
+                  _rtw_memcpy(psta->dot11tkiptxmickey.skey, &(param->u.crypt.key[16]), 8);
+                  _rtw_memcpy(psta->dot11tkiprxmickey.skey, &(param->u.crypt.key[24]), 8);
+              } else {
+                  RTW_WARN("TKIP key_len too short for MIC keys: %u\n", param->u.crypt.key_len);
+                  /* Optionally zero out the keys or handle error */
+                  _rtw_memset(psta->dot11tkiptxmickey.skey, 0, 8);
+                  _rtw_memset(psta->dot11tkiprxmickey.skey, 0, 8);
+              }
+              padapter->securitypriv.busetkipkey = _FALSE;
```
Please reformat code to comply with file standard

delpantheone replied:
Hello! Thanks for your heads up. I think i've fixed the code formatting now.

wiktorek140 commented on os_dep/linux/ioctl_cfg80211.c:
```diff
@@ -1586,8 +1593,14 @@ static int rtw_cfg80211_set_encryption(struct net_device *dev, struct ieee_param
 							(param->u.crypt.key_len > 16 ? 16 : param->u.crypt.key_len));
-						_rtw_memcpy(padapter->securitypriv.dot118021XGrptxmickey[param->u.crypt.idx].skey, &(param->u.crypt.key[16]), 8);
-						_rtw_memcpy(padapter->securitypriv.dot118021XGrprxmickey[param->u.crypt.idx].skey, &(param->u.crypt.key[24]), 8);
+            if (param->u.crypt.key_len >= 32) {
+                _rtw_memcpy(padapter->securitypriv.dot118021XGrptxmickey[param->u.crypt.idx].skey, &(param->u.crypt.key[16]), 8);
+                _rtw_memcpy(padapter->securitypriv.dot118021XGrprxmickey[param->u.crypt.idx].skey, &(param->u.crypt.key[24]), 8);
+            } else {
+                RTW_WARN("Group TKIP key_len too short for MIC keys: %u\n", param->u.crypt.key_len);
+                _rtw_memset(padapter->securitypriv.dot118021XGrptxmickey[param->u.crypt.idx].skey, 0, 8);
+                _rtw_memset(padapter->securitypriv.dot118021XGrprxmickey[param->u.crypt.idx].skey, 0, 8);
+            }
```
reformat code

delpantheone committed:
Add more boundary checks for setting MIC keys

By checking if param->u.crypt.key_len size is, at least 32 bytes, we
prevent UBSAN errors.

Why 32 bytes?

To be honest, i didn't even know before. But after some research i
figured out that it is from the IEEE 802.11i specification.

A TKIP key in 802.11 networks is typically 32 bytes long.

So, by checking param->u.crypt.key_len >= 32, we ensure the buffer has
enough space to safely copy both MIC keys

If the key is shorter, we log a warning and zero the keys to avoid
undefined behavior

delpantheone committed:
Formatted to comply with code standard

delpantheone commented on os_dep/linux/ioctl_cfg80211.c:
```diff
@@ -1586,8 +1593,14 @@ static int rtw_cfg80211_set_encryption(struct net_device *dev, struct ieee_param
 							(param->u.crypt.key_len > 16 ? 16 : param->u.crypt.key_len));
-						_rtw_memcpy(padapter->securitypriv.dot118021XGrptxmickey[param->u.crypt.idx].skey, &(param->u.crypt.key[16]), 8);
-						_rtw_memcpy(padapter->securitypriv.dot118021XGrprxmickey[param->u.crypt.idx].skey, &(param->u.crypt.key[24]), 8);
+            if (param->u.crypt.key_len >= 32) {
```
Let me explain.

But first, I need to point out that I am not a C developer. I am just a guy trying to fix a driver issue on my laptop.

This value was suggested by an AI while I was searching for a solution to the "array-out-of-bounds" error.

So I investigated why this magic number should be there.

After some research, I found out that this value is the typical size of a TKIP message.

It comes from the IEEE 802.11i specification.

After adding bounds checks, I was able to run the driver without any more "array-out-of-bounds" errors flooding the dmesg log.

This may not be the proper or most optimized way to do this kind of fix. But it worked for me and may work for others as well.
